internal/expr/bindtypes.go

```diff
@@ -75,63 +72,68 @@ func (pe *ParsedExpr) BindTypes(args ...any) (te *TypedExpr, err error) {
 		}
 	}
 
-	var sql bytes.Buffer
-
-	var inCount int
-	var outCount int
-
-	var outputs = make([]typeinfo.Member, 0)
-	var inputs = make([]typeinfo.Member, 0)
-
-	var typeMemberPresent = make(map[typeinfo.Member]bool)
+	// Bind types to each expression.
+	typedExprs := []typedExpression{}
+	outputMemberUsed := map[typeinfo.Member]bool{}
+	for _, expr := range *pe {
+		te, err := expr.bindTypes(ti)
+		if err != nil {
+			return nil, err
+		}
 
-	// Check and expand each query expr.
-	for _, expr := range pe.exprs {
-		switch e := expr.(type) {
-		case *inputExpr:
-			inLoc, err := bindInputTypes(ti, e)
+		if toe, ok := te.(*typedOutputExpr); ok {
+			err = toe.checkUsed(outputMemberUsed)
 			if err != nil {
 				return nil, err
 			}
-			sql.WriteString("@sqlair_" + strconv.Itoa(inCount))
-			inCount++
-			inputs = append(inputs, inLoc)
-		case *outputExpr:
-			outCols, typeMembers, err := bindOutputTypes(ti, e)
-			if err != nil {
-				return nil, err
-			}
-
-			for _, tm := range typeMembers {
-				if ok := typeMemberPresent[tm]; ok {
-					return nil, fmt.Errorf("member %q of type %q appears more than once in output expressions", tm.MemberName(), tm.OuterType().Name())
-				}
-				typeMemberPresent[tm] = true
-			}
-
-			for i, c := range outCols {
-				sql.WriteString(c.String())
-				sql.WriteString(" AS ")
-				sql.WriteString(markerName(outCount))
-				if i != len(outCols)-1 {
-					sql.WriteString(", ")
-				}
-				outCount++
-			}
-			outputs = append(outputs, typeMembers...)
-		case *bypass:
-			sql.WriteString(e.chunk)
-		default:
-			return nil, fmt.Errorf("internal error: unknown query expr type %T", expr)
 		}
+
+		typedExprs = append(typedExprs, te)
 	}
 
-	return &TypedExpr{inputs: inputs, outputs: outputs, sql: sql.String()}, nil
+	typedExpr := TypedExprs(typedExprs)
+	return &typedExpr, nil
 }
 
-// bindInputTypes binds the input expression to a type and returns the
-// typeMember represented by the expression.
-func bindInputTypes(ti typeNameToInfo, e *inputExpr) (tm typeinfo.Member, err error) {
+// expression represents a parsed node of the SQLair query's AST.
+type expression interface {
+	// String returns a text representation for debugging and testing purposes.
+	String() string
+
+	// bindTypes generates a typed expression from the query type information.
+	bindTypes(typeNameToInfo) (typedExpression, error)
+}
+
+// bypass represents part of the expression that we want to pass to the backend
+// database verbatim.
+type bypass struct {
+	chunk string
+}
+
+func (e *bypass) String() string {
+	return "Bypass[" + e.chunk + "]"
+}
+
+// bindTypes is part of the expression interface. bypass expressions have no
+// types so the same expression is returned.
+func (e *bypass) bindTypes(typeNameToInfo) (typedExpression, error) {
+	return e, nil
+}
+
+// inputExpr represents a named parameter that will be sent to the database
+// while performing the query.
```

Comment:
It does not represent a named parameter, that is what the struct is going to be used for. It represents a node in the AST into which the user can input a value.